Comprehensive STM32G0 driver

fabalthazar opened the pull request:
Fixes #819.

This PR provides a comprehensive STM32G0 target driver in a separate file. G0 has some divergences from L4/G4.
And may leave cleaner room in stm32l4.c for #707 and #458. Maybe...

Refer to this message for content details:
https://github.com/blacksphere/blackmagic/issues/819#issuecomment-780142079

Waiting for Dual-bank G0B/C release (in the ST's pipe) might be reasonable to fully test it.

UweBonnes commented:
Several requests.issues:
- I am somehow torn apart by the amount of additional flash (2.6 k) needed, but I am willing to accept.
- As this is new code, documenting steps forward and backward in the git history does not help other users to understand the code. So for the PR, squash into one file or discuss reasons not to do so
- Add real usage examples of the new features you introduced, either in the code, in the wiki or  where you think it fits.

UweBonnes commented:
Also if you do not mind: Tag new files that you created with a more liberal license than the GNU license, e.g. BSD.

fabalthazar commented:
Ouch 2.6k is enormous. I suspected but not that much.
I analysed the .text section usage at the function level (after making functions non static so the global footprint is even higher but it is relevant for proportions). Here is a summary:

- Fundamental functions related to probe, attach/detach, flash erase/write => 972 bytes
And for custom commands:
- erase mass/bank => 122 bytes
- option bytes => 938 bytes => way too much!

In the option bytes part, a single function _stm32g0_is_PCROP_active()_ occupies no less than 136 bytes by itself just for combinatorial computations to know whether the Proprietary Code Read Out Protection is active or not.

So I have two thoughts:

1. The primary idea to have a separate file (as discussed in linked #819) was ambitious. From this memory analysis, one can estimate that an additional STM32 driver will occupy at least ~1k of code for fundamental features. Here, this part brings STM32G0B/C dual-bank specificity so it can be managed in existing stm32l4.c at the price of additional if() or switch/case complexity;
2. For option bytes, I tried to implement it in a generic and flexible manner. So it could make sense if it is shared with other STM32 drivers. For example in a stm32_common.c file. If this stays isolated in G0 file, this surely is overkill. So apart from PCROP stuff, default bytes can be implemented in stm32l4.c too.

Generally speaking, have you thought about the concept of STM32 API? STM32 drivers often have the same patterns.
This may not be easy for flash management due to many chips specificities. Or at the cost of versioned functions like stm32_flash_write_v1, _v2 (that must be well documented!)... with many arguments to deal with registers.
Or maybe this concept can apply to option bytes, the way they are given on the command line, parsed, validated or not, written, erased, etc. Current STM32 drivers seam to have their own way of dealing with these bytes. And this could lead to even lower flash usage.

So so so... What I propose to transform this PR into. Only implement these little features/fix (from complete list in https://github.com/blacksphere/blackmagic/issues/819#issuecomment-780142079) in stm32l4.c:

- Support for STM32G05x/G06x;
- Really enable debug in low power modes (DBG peripheral clock enabled). Current L4 implementation may not work (it does not on G0 chips);
- Watchdogs (IWDG and WWDG) are stopped when the core is halted. I found it to be the only way to program a binary file with hosted blackmagic where the debug registers cannot be written. Already appeared in #117 ;
- Lock the Flash back after a flash operation to leave the device in a state as close as possible to its previous state;
- Properly clear EOP flag after a flash erasure/write as recommended;
- EMPTY bit cleared after a successful programming to allow rebooting in Main Flash memory without a necessary power off. Particularly useful in case of virgin programming in production where subsequent actions are done automatically, for instance;
    Option bytes:
- G0-specific default values. I did not reproduce an "option erase" command on G0 with current STM32L4 values so I cannot confirm whether it is the reason for my bricked Nucleo board;
- Option bytes support for G0 dual-bank devices.

The idea behind the _irreversible (enable|disable)_ was to eventually enable OTP area programming (as CMSIS-DAP). Perhaps makes sense in a shared common file too.

This proposal may assuredly limit a lot the memory blow up.

Your advise?

fabalthazar commented:
Let me know if you're interested in the concept of shared functions between STM32 (globally or limited to option bytes). In case I find time, I could prototype something.
Limited to option bytes may at least unify the option bytes programming CLI in the STM32 range. At the price of breaking compatibility with current users private scripts (needs doc in Wiki and release note).

UweBonnes commented on src/target/stm32g0.c:
```diff
@@ -0,0 +1,758 @@
+	uint32_t dbg_apb_fz1;
+};
+
+static struct saved_regs_s saved_regs = {0};
```
But that as piv in the target structure

fabalthazar replied:
Are you referring to "target_storage" field in the "target_s" structure?
I attempted to store a pointer to an allocated structure in target_storage thanks to the uintptr_t standard type but after tries and tries, it fails on 64 bits hosts as uintptr_t uses 8 bytes and target_storage is 4 bytes (segfault).
Other idea than static variables?

UweBonnes commented on src/target/stm32g0.c:
```diff
@@ -0,0 +1,758 @@
+};
+
+static struct saved_regs_s saved_regs = {0};
+static bool irreversible_enabled = false;
```
Why not as additional entity in saved_regs

fabalthazar replied:
Yes. I am waiting for knowing how to properly store "saved_regs".

UweBonnes commented on src/target/stm32g0.c:
```diff
@@ -0,0 +1,758 @@
+
+	target_mem_map_free(t);
+
+	size_t flash_size = (size_t)target_mem_read16(t, FLASH_MEMORY_SIZE);
```
I think it is best to announce the biggest flash size for that chip. Smaller flash sizes only means that that flash up to the real size is not tested or perhaps disabled. By using the largest size, we give the user control what to do with flash above the announced size. Do you have G0 devices with criplled flash? Can you test?

fabalthazar replied:
I got an STM32G031 (32 kiB) and I can confirm that I succeeded in writing above 0x8000!
I implemented the full 64 kiB announcement for single bank devices.

UweBonnes commented:
> Ouch 2.6k is enormous. I suspected but not that much.
That is in line with the other file:
   text    data     bss     dec     hex filename
   2772       0       0    2772     ad4 stm32f1.o
   4373       0       0    4373    1115 stm32f4.o
   2856       0      16    2872     b38 stm32g0.o
   2683       0       0    2683     a7b stm32h7.o
   4007       0       0    4007     fa7 stm32l0.o
   2428       0       0    2428     97c stm32l4.o
   2560       0       0    2560     a00 stm32l4.o (in master)

There are two aspects:
- Keeps things simple and straightforward. In that way, I like to apply your PR, perhaps squashed and with some usage example. Look at the comments too. I also would not mind a more liberal license like BSD, but that is your choice.
- When we hit the 120 k (ST-Bootloader firmware) limits, it will get complicated to decide what to include in F103 firmware build and what not

Reuse is fine, when good tested so few chances for colateral damage

My advise:
- Lets get this PR applied with few changes
- I appreciate if you work out your other proposal sometimes later

fabalthazar commented:
Work still in progress.

I just put an hand on a fresh STM32G0B1 (512 kiB) dual-bank device! They are out.
I fixed a little thing in page erasure.
I confirm pages are organized a little bit differently than L4.

I removed the PCROP-related code that did not make sense until BMP does not support automatic re-scanning and attaching. I kept it simple(r).

I squashed the mass erase functions into a unique "erase_mass [1|2]" one to deal with the whole flash memory or specified bank number.

You might notice a little flash usage reduction.

I tested all this quite heavily.

Still to do:
- target_storage following your guidelines and help;
- Squash commits;
- Licence (I don't mind a more liberal one, I will have a look at this point);
- I noted your remarks concerning documentation too. I intend to populated the Wiki if its OK. I find it a bit skinny concerning target-specific usage. Perhaps I will clone the Wiki to work on it with a clear head.

fabalthazar commented:
Almost finished.

May you review `static struct stm32g0_priv_s stm32g0_private_storage` I am using to store target specific data?
As I am storing several data (uint32s + bool), I have not found a way to use the "target_storage" field in "target_s" struct since it is uint32_t.

Any idea other than current "static"?

UweBonnes commented:
What about extending "target_storage"? Instead of "uint32_t target_storage" "uint32_t target_storage[4]"

fabalthazar commented:
OK. Since it is common to all drivers, uint32_t *target_storage sounds more flexible. But needs memory allocation... source of leakage.
Or statically reserved table as you suggest but the table size is only allowed to increase in the future.
Feasible with few alignments is stm32h7 and l4.
And the drivers are responsible of having the knowledge of indices in the table.

UweBonnes commented:
"source of leakage" It  is allocated with the target structure and freed with the target structure. That happens already now when *target_storage" is static allocated. "drivers are responsible of having the knowledge of indices in the table" that happens with all those "priv" structures already.

But don't bang your head to much about these problems. There is a concept of "good enough"  and I think your PR is already there (when sensible squashed). But if you like to think further, you are welcome.

UweBonnes commented:
But what has to be kept in mind that something like 
"static struct saved_regs_s saved_regs = {0};"
will make problems when in the SWJ chain there are several targets of the same kind.

fabalthazar commented:
Rebased and squashed.

Target data is now allocated in target_storage and properly freed.

Small single-bank chips are announced with their max flash capacity ("overflashing" tested on STM32G031) ;
Dual bank devices get erased and flashed smoothly, including mass erase of bank 1 or 2 (tested on STM32G0B1) ;
Option bytes and full debugging experience tested on STM32G031, G070, G0B1.

Memory footprint: 2kB.

I am feeling nice with this work. Please have a look and criticize.

UweBonnes commented:
Thanks!
